pkg/util/cgroups/reader.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2016-present Datadog, Inc.
+
+package cgroups
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+)
+
+// Reader is the main interface to scrape data from cgroups
+type Reader struct {
+	hostPrefix             string
+	procPath               string
+	cgroupVersion          int
+	cgroupV1BaseController string
+	readerFilter           ReaderFilter
+	impl                   readerImpl
+
+	cgroups         map[string]Cgroup
+	scrapeTimestmap time.Time
+	cacheValidFor   time.Duration
+}
+
+type readerImpl interface {
+	parseCgroups() (map[string]Cgroup, error)
+}
+
+type ReaderFilter func(path, name string) (string, error)
+
+// DefaultFilter matches all cgroup folders and use folder name as identifier
+func DefaultFilter(path, name string) (string, error) {
+	return path, nil
+}
+
+var containerRegexp = regexp.MustCompile("([0-9a-f]{64}|[0-9a-f]{8}(-[0-9a-f]{4}){4})")
+
+// ContainerFilter returns a filter that will match cgroup folders containing a container id
+func ContainerFilter(path, name string) (string, error) {
+	match := containerRegexp.FindString(name)
+
+	// With systemd cgroup driver, there may be a `.mount` cgroup on top of the normal one
+	// While existing, no process is attached to it and thus holds no stats
+	if match != "" {
+		if strings.HasSuffix(name, ".mount") {
+			return "", nil
+		}
+
+		return match, nil
+	}
+
+	return "", nil
+}
+
+type ReaderOption func(*Reader)
+
+// WithHostPrefix sets where hosts path are mounted (if not running on-host)
+func WithHostPrefix(hostPrefix string) ReaderOption {
+	return func(r *Reader) {
+		r.hostPrefix = hostPrefix
+	}
+}
+
+// WithProcPath sets where /proc is currenty mounted.
+// If set, hostPrefix is not added to this path.
+// Default to `$hostPrefix/proc` if empty.
+func WithProcPath(fullPath string) ReaderOption {
+	return func(r *Reader) {
+		r.procPath = fullPath
+	}
+}
+
+// WithReaderFilter sets the filter used to select interesting cgroup folders
+// and provides an identifier for them.
+func WithReaderFilter(rf ReaderFilter) ReaderOption {
+	return func(r *Reader) {
+		r.readerFilter = rf
+	}
+}
+
+// WithCgroupV1BaseController sets which controller is used to select cgroups
+// it then assumes that, if being, used other controllers uses the same relative path.
+// Default to "memory" if not set.
+func WithCgroupV1BaseController(controller string) ReaderOption {
+	return func(r *Reader) {
+		r.cgroupV1BaseController = controller
+	}
+}
+
+// WithCgroupPathsCache sets the duration during which the discovered cgroup paths are considered valid.
+// Note that only the paths are cached, not the statistics.
+// Default to 0 if not set (no cache).
+func WithCgroupPathsCache(validFor time.Duration) ReaderOption {
+	return func(r *Reader) {
+		r.cacheValidFor = validFor
+	}
+}
+
+// NewReader returns a new cgroup reader with given options
+func NewReader(opts ...ReaderOption) (*Reader, error) {
+	r := Reader{}
```

Comment:
```suggestion
	r := &Reader{}
```
to directly define a pointer